### PR TITLE
Move all C++ impl. to ./cpp and remove Makefiles

### PR DESCRIPTION
**Pull request by tom91136**
This PR moves all C++ source files to the `./cpp` directory.
The CI and documentation has been updated to reflect this.

Motivation:
 * GitHub's README rendering is now very far down as the file list grows with new implementations. 
 * We might be adding new implementations in different programming languages (see #78, #95), having implementations grouped by language at the top-level helps with organisation.

### CODE REVIEW

**Comment by andreipoe** [comment]
How about `src` instead of `cpp`?

**Comment by tomdeakin** [comment]
This makes sense. I'd prefer `src/` too.

I wonder if we want to extend this and have `src/openmp`, `src/sycl`, etc...

**Comment by tom91136** [comment]
So with `src`, we get something like:
```

**Comment by tomdeakin** [comment]
Lets go with that structure as a first pass.

What do we think about having each model in a subdirectory inside `src/`?

**Comment by andreipoe** [comment]
That may be a good idea. If we then have all the Makefile use standard names, we could compile with just `make -C OpenMP`.

**Comment by tom91136** [comment]
Yep, I'm doing that now, so we'll end up with a `Makefile` in each `/src/<impl>/`

**Comment by tom91136** [comment]
Ready for review, I've renamed the `<model>.cmake` files to just `model.cmake` so as to not duplicate the same information.

File: .gitignore [hunk 1]
```diff
@@ -1,30 +1,8 @@
...
-*.o
-*.bc
-*.sycl
-*.tar
-*.gz
-*.a
-
-KokkosCore_config.*
-
-.DS_Store
```
Comment (tomdeakin) [inline .gitignore:21]: Top level .gitignore should include this `.DS_Store`

File: .gitignore [hunk 2]
```diff
@@ -1,30 +1,8 @@
...
-
-Makefile
 
 build/
 cmake-build-*/
 CMakeFiles/
 .idea/
 .vscode/
-.directory
```
Comment (tomdeakin) [inline .gitignore:30]: What is this for?

File: legacy/Makefile [hunk 3]
```diff
@@ -14,7 +14,8 @@ endif
 
 
 hc-stream: ../main.cpp HCStream.cpp
```
Comment (tomdeakin) [inline legacy/Makefile:16]: This file rename looks wrong. There might be multiple implementations placed in `legacy/` so we need to keep the Makefile name.

File: src/ci-test-compile.sh [hunk 4]
```diff
@@ -86,38 +86,38 @@ run_build() {
 }
```
Comment (tomdeakin) [inline src/ci-test-compile.sh:86]: The CI stuff isn't part of the BabelStream source - can this live in the top level directory?

File: src/ocl/Makefile [hunk 5]
```diff
@@ -30,8 +30,8 @@ else
   LIBS = -lOpenCL
 endif
 
-ocl-stream: main.cpp OCLStream.cpp
-	$(CXX) $(CXXFLAGS) -DOCL $^ $(EXTRA_FLAGS) $(LIBS) -o $@
+ocl-stream: ../main.cpp OCLStream.cpp
+	$(CXX) $(CXXFLAGS) -DOCL $^ $(EXTRA_FLAGS) $(LIBS) -o $@ -I. -I..
```
Comment (tomdeakin) [inline src/ocl/Makefile:34]: The local copy of the OpenCL headers in `CL` have been put in `src/CL` but they should go in `src/ocl/CL` instead.

File: src/omp/Makefile [hunk 6]
```diff
@@ -91,8 +91,8 @@ endif
 
 OMP = $(OMP_$(COMPILER)_$(TARGET))
 
-omp-stream: main.cpp OMPStream.cpp
-	$(CXX) $(CXXFLAGS) -DOMP $^ $(OMP) $(EXTRA_FLAGS) -o $@
+omp-stream: ../main.cpp OMPStream.cpp
+	$(CXX) $(CXXFLAGS) -DOMP  $^ $(OMP) $(EXTRA_FLAGS) -o $@ -I. -I..
```
Comment (tomdeakin) [inline src/omp/Makefile:95]: Do we need to specify `-I.`? This comment applies to all the Makefiles where appropriate.

File: src/std/Makefile [hunk 7]
```diff
@@ -6,8 +6,8 @@
 CXXFLAGS=-O3 -std=c++17 -stdpar -DSTD
 STD_CXX=nvc++
 
-std-stream: main.cpp STDStream.cpp
-	$(STD_CXX) $(CXXFLAGS) $^ $(EXTRA_FLAGS) -o $@
+std-stream: ../main.cpp STDStream.cpp
+	$(STD_CXX) $(CXXFLAGS) $^ $(EXTRA_FLAGS) -o $@ -I. -I..
```
Comment (tomdeakin) [inline src/std/Makefile:10]: `-I.` shouldn't be required here as we're not including any local headers that aren't already included with `#include "name.h"`

File: src/std20/Makefile [hunk 8]
```diff
@@ -17,8 +17,8 @@ FLAGS_GNU = -O3 -std=c++2a -march=native
 CXXFLAGS = $(FLAGS_$(COMPILER))
 
 
-std20-stream: main.cpp STD20Stream.cpp
-	$(CXX) -DSTD20 $(CXXFLAGS) $^ $(EXTRA_FLAGS) -o $@
+std20-stream: ../main.cpp STD20Stream.cpp
+	$(CXX) -DSTD20 $(CXXFLAGS) $^ $(EXTRA_FLAGS) -o $@ -I. -I..
```
Comment (tomdeakin) [inline src/std20/Makefile:21]: No need to `-I.`

File: src/sycl/Makefile [hunk 9]
```diff
@@ -73,8 +73,8 @@ SYCL_LINK_FLAGS = $(SYCL_$(COMPILER)_LINK_FLAGS)
 SYCL_INCLUDE = $(SYCL_$(COMPILER)_INCLUDE)
 
 # only ComputeCpp generates .sycl files which is a bit odd to deal with so we opted to compile everything together
-sycl-stream: main.cpp SYCLStream.cpp
-	$(SYCL_SYCLCXX) $(SYCL_SYCLFLAGS) $(SYCL_FLAGS) $(SYCL_INCLUDE) -DSYCL $(EXTRA_FLAGS) $(SYCL_LINK_FLAGS) $^ -o $@
+sycl-stream: ../main.cpp SYCLStream.cpp
+	$(SYCL_SYCLCXX) $(SYCL_SYCLFLAGS) $(SYCL_FLAGS) $(SYCL_INCLUDE) -DSYCL $(EXTRA_FLAGS) $(SYCL_LINK_FLAGS) $^ -o $@ -I. -I..
```
Comment (tomdeakin) [inline src/sycl/Makefile:77]: No need to `-I.`

File: src/sycl/SYCLStream.h [hunk 10]
```diff
@@ -11,6 +11,15 @@
...
 
+#include "CL/opencl.h"
+
+// XXX Intel's SYCL impl. needs CL_MEM_CHANNEL_INTEL which is provided in dpcpp's include dir
+// however, depending the system configuration, the system CL header sometimes takes precedence
+// we only really need this macro to refer to the extension so this is probably OK
+#ifndef CL_MEM_CHANNEL_INTEL
+#define CL_MEM_CHANNEL_INTEL            0x4213
+#endif
+
```
Comment (tomdeakin) [inline src/sycl/SYCLStream.h:22]: Changes like this should be done in the build system, not in the source code.
Even then, only if they are strictly required. If the CMake build finds the wrong header, then please change it so it requires the correct header.
I do not want to include a define copied from a vendor header in this repo!

File: src/sycl/model.cmake [hunk 11]
```diff
@@ -48,7 +48,7 @@ macro(setup)
         set(ComputeCpp_DIR ${SYCL_COMPILER_DIR})
 
         # don't point to the CL dir as the imports already have the CL prefix
-        set(OpenCL_INCLUDE_DIR "${CMAKE_SOURCE_DIR}/CL")
+        set(OpenCL_INCLUDE_DIR "${CMAKE_SOURCE_DIR}")
```
Comment (tomdeakin) [inline src/sycl/model.cmake:51]: Likely need to change this after removing the include in the source.

**Review by tomdeakin — CHANGES_REQUESTED** [review]

**Comment by tomdeakin** [comment]
As we're about to merge in Julia, Rust, etc. I wonder if those should go into a `standalone` directory. Thoughts?

File: .gitignore (hunk 2 above)
Comment (tom91136) [reply .gitignore:30]: KDE's Dolphin, and maybe some other file managers leave that, similar to `.DS_store`

**Comment by tom91136** [comment]
Yeah, I'm OK with that. 
If we'll only ever have a few standalones ones then I think they can live at the top-level but it looks like we might add a few more later.
Not sure if reaching PRK levels of language/framework coverage is a future milestone.

I think we should merge all the standalones PRs (Julia, Rust, and maybe Scala) first and then I can figure out how to make this PR work with the CI.

**Comment by tomdeakin** [comment]
Will resolve #110 on merge

**Comment by tomdeakin** [comment]
This PR will be updated to do the following:
- Move all implementations to `./src/<model>/{*.cpp, ...}`
- Move standalone implementations to e.g. `./src/Scala/...`
- Remove Makefile support in favour of CMake only
- Update the CI to work with new directory structure
- Download CL headers on demand where required, and remove otherwise. Retain `cl.hpp` in new `src/ocl/`. This will resolve #109.

**Comment by tom91136** [comment]
@tomdeakin CI is green now, ready for merge

File: src/sycl/SYCLStream.h [hunk 12]
```diff
@@ -11,6 +11,15 @@
 
 #include "Stream.h"
 
+#include "CL/opencl.h"
```
[the remark this answers is not in the document]
Comment (tom91136) [reply src/sycl/SYCLStream.h]: Removed

File: src/sycl/SYCLStream.h (hunk 10 above)
Comment (tom91136) [reply src/sycl/SYCLStream.h]: Removed

**Pull request merged**